src/application_input.hpp

```diff
@@ -37,7 +37,10 @@ struct ApplicationInput {
 
   // Mesh functions
   std::function<void(ParameterInput *)> InitUserMeshData = nullptr;
-  std::function<void()> MeshUserWorkInLoop = nullptr;
+  std::function<void(Mesh *, ParameterInput *, SimTime const &)>
+      PreStepMeshUserWorkInLoop = nullptr;
+  std::function<void(Mesh *, ParameterInput *, SimTime const &)>
+      PostStepMeshUserWorkInLoop = nullptr;
```

Comment:
My thought earlier was maybe there should be an application-wide `OutputDiagnostics` which could go here in `ApplicationInput` as well as the `UserWorkInLoop` stuff.

This is optional though. I think having package-level diagnostics and working application-level work in loop is already a major improvement.

Reply:
Ok, I have added `Pre`/`PostStepDiagnostics` for the application.  The mesh has default implementation that cycle each package and call their `Pre`/`PostStepDiagnostics` functions.